### PR TITLE
[Hub Generated] Review request for Microsoft.ContainerRegistry to add version stable/2023-07-01

### PR DESCRIPTION
**Pull request by AntonioVT**
This is a PR generated at [OpenAPI Hub](https://aka.ms/openapihub). You can view your [work branch via this link](https://openapihub.azure-devex-tools.com/branch/Azure/azure-rest-api-specs/AntonioVT-containerregistry-Microsoft.ContainerRegistry-2023-07-01...main/).
# ARM (Control Plane) API Specification Update Pull Request 

## Purpose of this PR

What's the purpose of this PR? Check all that apply. This is **mandatory**!

  - [x] New API version. (Such PR should have been generated with [OpenAPI Hub](https://aka.ms/openapihub), per [this wiki doc](https://dev.azure.com/azure-sdk/internal/_wiki/wikis/internal.wiki/208/OpenAPI-Hub-Adding-new-API-version).)
  - [ ] Update existing version for a new feature. (This is applicable only when you are revising a private preview API version.)
  - [ ] Update existing version to fix swagger quality issues in S360.
  - [ ] Other, please clarify:
    - _edit this with your clarification_

## Due diligence checklist

To merge this PR, you **must** go through the following checklist and confirm you understood 
and followed the instructions by checking all the boxes:

- [x] I have reviewed the general guidance on the spec PR review process: https://aka.ms/specprreview.
- [x] I confirm this PR is modifying Azure Resource Manager (ARM) related specifications, and not data-plane related specifications.
- [x] I commit to follow the [Breaking Change Policy](https://aka.ms/AzBreakingChangesPolicy).
- [x] I have reviewed following [Resource Provider guidelines](https://aka.ms/rpguidelines), including
  [ARM resource provider contract](https://github.com/Azure/azure-resource-manager-rpc) and
  [REST guidelines](https://github.com/microsoft/api-guidelines/blob/vNext/azure/Guidelines.md) (estimated time: 4 hours).  
  I understand this is required before I can request review from an ARM API Review board.

### ARM API changes review

- If you want for the ARM team to review this PR, you must add the `ARMReview` label. 
- The automation may automatically add the `ARMReview` label, if appropriate.  
  If this happens, proceed according to guidance given in GitHub comments also added by the automation.

### Breaking change review

If you have any breaking changes as defined in the [Breaking Change Policy](https://aka.ms/AzBreakingChangesPolicy/), 
follow the process outlined in the [High-level Breaking Change Process doc](https://aka.ms/breakingchangesprocess#high-level-breaking-change-process).
      
## Getting help

- For guidance on fixing this PR CI check failures, see the hyperlinks provided in given failure 
  and https://aka.ms/ci-fix.
- For additional help, see https://aka.ms/azsdk/support/spectools.

### CODE REVIEW

**Comment by AntonioVT** [comment]
Hi Swagger/ARM review team. The PR includes the following changes:

- Add Cache Rules and Credential Sets resource types/operations to GA version
- Fix properties that were meant to be readonly (IdentityProperties and UserIdentityProperties properties)

**Comment by AntonioVT** [comment]
Regarding the CI issues, I think they are false positives:

Avocado: Issues are related to older api versions, this should be fixed in a separate PR
Semantic Validation, Model Validation, LintDiff: These are passing ️️✔️

cc @rkmanda @msyyc @tadelesh

**Comment by AntonioVT** [comment]
Hi @qiaozha, there's currently a CI failure in the SDK Azure-sdk-for-go pipeline. There's a duplicate schema "ErrorResponse" with certain differences between our containerregistry.json and containerregistry_build.json. 

The containerregistry.json is now using the ARM ErrorResponse definition from common-types/v5/types.json. However, since this change is not present in containerregistry_build.json, it's being considered a breaking change. I believe we can add a directive to use a specific schema and fix the CI.

The differences between our custom ErrorResponse and types.json ErrorResponse definitions can be found here:
[ACR ResponseError definitions (local definition, old version)](https://microsoft-my.sharepoint.com/:u:/p/anvarga/EVpkhvopHvdOmVwYhXQ6yDgBFB8WSIZUEua2Vx_1T4EcgQ?e=qVymP8) 
[CommonTypes ResponseError definitions (external definition, current version)](https://microsoft-my.sharepoint.com/:u:/p/anvarga/EdrTMsePT8BIqLNAf011p3wBZ6auiPfBFPLV34ihbzynfg?e=nFDAPX)

In summary, there are new properties which are optional, they are not decorated with the "required" swagger attribute (`details` and `additionalInfo` properties under `ErrorDetail`).

**Comment by konrad-jamrozik** [comment]
> Avocado CI false positive, [CI results here](https://github.com/Azure/azure-rest-api-specs/pull/24771/checks?check_run_id=15442088184), analysis below:
> 
> I noticed two issues in the Avocado CI that are being marked as errors, however I don't think this is accurate. The errors are "MISSING_APIS_IN_DEFAULT_TAG".
> 
> In Summary, the errors are related to paths that are not supported in this stable version because we either have stopped supporting those paths long time ago, or they are preview feature paths.
> 
> 1. The API path "/subscriptions/{}/resourceGroups/{}/providers/Microsoft.ContainerRegistry/registries/{}/listPolicies" and ".../updatePolicies" is no longer supported.
>    
>    * The last stable API version that supports it is **2017**-10-01. The next GA version is 2019-05-01 and doesn't include that path anymore. The last stable version is **2022-12-01** and it also doesn't include the paths mentioned.
> 2. The API path "/subscriptions/{}/resourceGroups/{}/providers/Microsoft.ContainerRegistry/registries/{}/exportPipeline*", "importPipeline*", "pipelineRuns*", etc are PREVIEW path operations.
>    
>    * These features are not ready to be in stable version. These errors are being flagged from the 2023-06-01-preview version. This comparison should be done with another preview version, but not with the current stable version.

Sorry for the issue, and thank you for the explanation.

Based on your description, you ran into this limitation:
- https://github.com/Azure/azure-sdk-tools/issues/6627
 
I just updated [the doc](https://github.com/Azure/avocado/blob/main/README.md#missing_apis_in_default_tag) to mention how to handle your cases.

I will add now `Approved-Avocado` to suppress the Avocado failure.

**Comment by raych1** [comment]
> Hi @raych1, there's currently a CI failure in the SDK Azure-sdk-for-go pipeline. There's a duplicate schema "ErrorResponse" with certain differences between our containerregistry.json and containerregistry_build.json.
> 
> The containerregistry.json is now using the ARM ErrorResponse definition from common-types/v5/types.json. However, since this change is not present in containerregistry_build.json, it's being considered a breaking change. I believe we can add a directive to use a specific schema and fix the CI. This was done in our previous preview release [here](https://github.com/Azure/azure-rest-api-specs/pull/24426#discussion_r1266088164).
> 
> The differences between our custom ErrorResponse and types.json ErrorResponse definitions can be found here: [ACR ResponseError definitions (local definition, old version)](https://microsoft-my.sharepoint.com/:u:/p/anvarga/EVpkhvopHvdOmVwYhXQ6yDgBFB8WSIZUEua2Vx_1T4EcgQ?e=qVymP8) [CommonTypes ResponseError definitions (external definition, current version)](https://microsoft-my.sharepoint.com/:u:/p/anvarga/EdrTMsePT8BIqLNAf011p3wBZ6auiPfBFPLV34ihbzynfg?e=nFDAPX)
> 
> In summary, there are new properties which are optional, they are not decorated with the "required" swagger attribute (`details` and `additionalInfo` properties under `ErrorDetail`).

@AntonioVT , we can add directive to bypass this issue. However, can you use the same referenced `ErrorResponse` or use a different name in next new api version?

@Alancere , can you help add the directive for go readme?

**Comment by Alancere** [comment]
Added directive to readme.go.md

**Comment by AntonioVT** [comment]
> > Hi @raych1, there's currently a CI failure in the SDK Azure-sdk-for-go pipeline. There's a duplicate schema "ErrorResponse" with certain differences between our containerregistry.json and containerregistry_build.json.
> > The containerregistry.json is now using the ARM ErrorResponse definition from common-types/v5/types.json. However, since this change is not present in containerregistry_build.json, it's being considered a breaking change. I believe we can add a directive to use a specific schema and fix the CI. This was done in our previous preview release [here](https://github.com/Azure/azure-rest-api-specs/pull/24426#discussion_r1266088164).
> > The differences between our custom ErrorResponse and types.json ErrorResponse definitions can be found here: [ACR ResponseError definitions (local definition, old version)](https://microsoft-my.sharepoint.com/:u:/p/anvarga/EVpkhvopHvdOmVwYhXQ6yDgBFB8WSIZUEua2Vx_1T4EcgQ?e=qVymP8) [CommonTypes ResponseError definitions (external definition, current version)](https://microsoft-my.sharepoint.com/:u:/p/anvarga/EdrTMsePT8BIqLNAf011p3wBZ6auiPfBFPLV34ihbzynfg?e=nFDAPX)
> > In summary, there are new properties which are optional, they are not decorated with the "required" swagger attribute (`details` and `additionalInfo` properties under `ErrorDetail`).
> 
> @AntonioVT , we can add directive to bypass this issue. However, can you use the same referenced `ErrorResponse` or use a different name in next new api version?
> 
> @Alancere , can you help add the directive for go readme?

Yes, we will reference the ARM defined ErrorResponsewhen when we release a new containerregsitry_build API.

**Comment by zizw123** [comment]
>           "201": {

Should use 202 for async PATCH, and also use location header, see [https://github.com/Azure/azure-resource-manager-rpc/blob/master/v1.0/async-api-reference.md#updating-using-patch](https://github.com/Azure/azure-resource-manager-rpc/blob/master/v1.0/async-api-reference.md#updating-using-patch)

---
Refers to: specification/containerregistry/resource-manager/Microsoft.ContainerRegistry/stable/2023-07-01/containerregistry.json:268 in 2f24e4f. [](commit_id = 2f24e4fd0275eb41cc31384478a4ed569a7a0b50, deletion_comment = False)

**Comment by zizw123** [comment]
>           "201": {

Same for this

---
Refers to: specification/containerregistry/resource-manager/Microsoft.ContainerRegistry/stable/2023-07-01/containerregistry.json:268 in 2f24e4f. [](commit_id = 2f24e4fd0275eb41cc31384478a4ed569a7a0b50, deletion_comment = False)

**Comment by AntonioVT** [comment]
> > ```
> >       "201": {
> > ```
> 
> Should use 202 for async PATCH, and also use location header, see https://github.com/Azure/azure-resource-manager-rpc/blob/master/v1.0/async-api-reference.md#updating-using-patch
> 
> Refers to: specification/containerregistry/resource-manager/Microsoft.ContainerRegistry/stable/2023-07-01/containerregistry.json:268 in [2f24e4f](https://github.com/Azure/azure-rest-api-specs/commit/2f24e4fd0275eb41cc31384478a4ed569a7a0b50). [](commit_id = [2f24e4f](https://github.com/Azure/azure-rest-api-specs/commit/2f24e4fd0275eb41cc31384478a4ed569a7a0b50), deletion_comment = False)

CacheRules and CredentialSet PATCH behavior is the same as the rest of ACR proxy resources. We had a discussion with @visingla-ms when we released this feature (in 2023-01-01 preview API) and determined that it was OK to keep the current behavior in our APIs. Could we keep the current behavior?

Thread references: 
https://github.com/Azure/azure-rest-api-specs/pull/21727#issuecomment-1335998198
https://github.com/Azure/azure-rest-api-specs/pull/21727#issuecomment-1336056854

**Comment by AntonioVT** [comment]
@ms-zhenhua I have synced offline with @zizw123 regarding our current PATCH behavior across all of our APIs. Our backend is already returning the 201. We have a work item to migrate our PATCH operations across all resources, dev ops link below. I have removed the ARMChangesRequested label. Could you help us resume the PR review please? 

[24778096 [RP] Migrate RP resources PATCH operations to follow ARM RPC](https://msazure.visualstudio.com/AzureContainerRegistry/_workitems/edit/24778096)

**Comment by ms-zhenhua** [comment]
>           },

LRO-Delete does not need 200

---
Refers to: specification/containerregistry/resource-manager/Microsoft.ContainerRegistry/stable/2023-07-01/containerregistry.json:200 in 2f24e4f. [](commit_id = 2f24e4fd0275eb41cc31384478a4ed569a7a0b50, deletion_comment = False)

**Comment by ms-zhenhua** [comment]
>           "201": {

resolved with [https://msazure.visualstudio.com/AzureContainerRegistry/_workitems/edit/24778096](https://msazure.visualstudio.com/AzureContainerRegistry/_workitems/edit/24778096)#

---
In reply to: [1666239876](https://github.com/Azure/azure-rest-api-specs/pull/24771#issuecomment-1666239876) [](http://example.com/codeflow?ancestors=1666239876)

---
Refers to: specification/containerregistry/resource-manager/Microsoft.ContainerRegistry/stable/2023-07-01/containerregistry.json:268 in 2f24e4f. [](commit_id = 2f24e4fd0275eb41cc31384478a4ed569a7a0b50, deletion_comment = False)

**Comment by ms-zhenhua** [comment]
>           "201": {

resolved with [https://msazure.visualstudio.com/AzureContainerRegistry/_workitems/edit/24778096](https://msazure.visualstudio.com/AzureContainerRegistry/_workitems/edit/24778096)#

---
In reply to: [1666239285](https://github.com/Azure/azure-rest-api-specs/pull/24771#issuecomment-1666239285) [](http://example.com/codeflow?ancestors=1666239285)

---
Refers to: specification/containerregistry/resource-manager/Microsoft.ContainerRegistry/stable/2023-07-01/containerregistry.json:268 in 2f24e4f. [](commit_id = 2f24e4fd0275eb41cc31384478a4ed569a7a0b50, deletion_comment = False)

**Comment by ms-zhenhua** [comment]
>           },

LRO-Delete does not need 200

---
Refers to: specification/containerregistry/resource-manager/Microsoft.ContainerRegistry/stable/2023-07-01/containerregistry.json:479 in 2f24e4f. [](commit_id = 2f24e4fd0275eb41cc31384478a4ed569a7a0b50, deletion_comment = False)

**Comment by ms-zhenhua** [comment]
>     "IdentityProperties": {

consider using common type:
[https://github.com/Azure/azure-rest-api-specs/blob/ad997e99eccc15b7ab4cd66ae3f1f9534a1e2628/specification/common-types/resource-management/v5/managedidentity.json#L50](https://github.com/Azure/azure-rest-api-specs/blob/ad997e99eccc15b7ab4cd66ae3f1f9534a1e2628/specification/common-types/resource-management/v5/managedidentity.json#L50)

---
Refers to: specification/containerregistry/resource-manager/Microsoft.ContainerRegistry/stable/2023-07-01/containerregistry.json:2743 in 2f24e4f. [](commit_id = 2f24e4fd0275eb41cc31384478a4ed569a7a0b50, deletion_comment = False)

**Comment by ms-zhenhua** [comment]
>         "provisioningState": {

duplicated with that in "CacheRuleProperties". Consider defining a common model

---
Refers to: specification/containerregistry/resource-manager/Microsoft.ContainerRegistry/stable/2023-07-01/containerregistry.json:2804 in 2f24e4f. [](commit_id = 2f24e4fd0275eb41cc31384478a4ed569a7a0b50, deletion_comment = False)

**Comment by ms-zhenhua** [comment]
>     "UserIdentityProperties": {

consider using common type:
[https://github.com/Azure/azure-rest-api-specs/blob/ad997e99eccc15b7ab4cd66ae3f1f9534a1e2628/specification/common-types/resource-management/v5/managedidentity.json#L18](https://github.com/Azure/azure-rest-api-specs/blob/ad997e99eccc15b7ab4cd66ae3f1f9534a1e2628/specification/common-types/resource-management/v5/managedidentity.json#L18)

---
Refers to: specification/containerregistry/resource-manager/Microsoft.ContainerRegistry/stable/2023-07-01/containerregistry.json:2823 in 2f24e4f. [](commit_id = 2f24e4fd0275eb41cc31384478a4ed569a7a0b50, deletion_comment = False)

**Comment by ms-zhenhua** [comment]
>     "AuthCredential": {

please double check to use x-ms-secret for credential properties
[https://github.com/Azure/autorest/blob/main/docs/extensions/readme.md#x-ms-secret](https://github.com/Azure/autorest/blob/main/docs/extensions/readme.md#x-ms-secret)

---
Refers to: specification/containerregistry/resource-manager/Microsoft.ContainerRegistry/stable/2023-07-01/containerregistry.json:2883 in 2f24e4f. [](commit_id = 2f24e4fd0275eb41cc31384478a4ed569a7a0b50, deletion_comment = False)

**Comment by AntonioVT** [comment]
Hi @ms-zhenhua, regarding
- LRO-Delete, I have deleted the 200 status response. I have confirmed our backend doesn't support it, so it was unnecessary to have it in swagger. I have created [WI [RP] [Swagger] Review and Remove 200 Status code in LRO-Delete operations in registry/child resources](https://msazure.visualstudio.com/AzureContainerRegistry/_workitems/edit/24802955) so we review/fix instances where this is happening for next api version. 
- Duplicate properties across different properties, I have created [[RP] [Swagger] Review and update/define common models](https://msazure.visualstudio.com/AzureContainerRegistry/_workitems/edit/24803066). There are a lot of instances where this is happening, and this should be included in our swagger generation tooling. Do you have a concern with keeping it for this PR as is?
- x-ms-secret, is not applicable to our scenario. The contents are a keyvault URI, not sensitive data. We also had this observation in our initial PR (preview) [Thread here (last paragraph)](https://github.com/Azure/azure-rest-api-specs/pull/21727#issuecomment-1335998198).

Thanks!

**Comment by ms-zhenhua** [comment]
>           "201": {

LRO-Patch should use 200 and 202

---
Refers to: specification/containerregistry/resource-manager/Microsoft.ContainerRegistry/stable/2023-07-01/containerregistry.json:265 in 05a2ef1. [](commit_id = 05a2ef1a79e67477beb5a3e184559379849655ad, deletion_comment = False)

**Comment by ms-zhenhua** [comment]
>           "201": {

same as above

---
Refers to: specification/containerregistry/resource-manager/Microsoft.ContainerRegistry/stable/2023-07-01/containerregistry.json:541 in 05a2ef1. [](commit_id = 05a2ef1a79e67477beb5a3e184559379849655ad, deletion_comment = False)

**Comment by AntonioVT** [comment]
@ms-zhenhua Regarding LRO-Patch, my understanding is that this has been resolved with WI https://msazure.visualstudio.com/AzureContainerRegistry/_workitems/edit/24778096#. 

Reference: 
https://github.com/Azure/azure-rest-api-specs/pull/24771#issuecomment-1668814542
https://github.com/Azure/azure-rest-api-specs/pull/24771#issuecomment-1666239876

**Comment by kazrael2119** [comment]
this js sdk breaking is caused by this pr: change ErrorResponse from manually definition to common-type

**Comment by raych1** [comment]
@AntonioVT , this PR contains a bunch of removals comparing to `2023-06-01-preview` version, these changes would cause breaking change. Did you go thru them with breaking change review board? I want to call out these removals are not reported in above `cross version breaking change` CI check because of a tool bug. 

//CC: @mikekistler , @JeffreyRichter

**Comment by AntonioVT** [comment]
Hi @ms-zhenhua, @raych1, it seems that we have signoff from ARM and SDKs. Is there any approval left? We would like to start the ARM manifest rollout.

**Review by rkmanda — APPROVED** [review]
